Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,849 @@
-        $Error.Clear()
-        Start-Sleep -Seconds 2
-        exit
-    }
-    $dateTime = Get-Date
-    $installPath = (Get-ItemProperty -Path Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\ExchangeServer\v15\Setup -ErrorAction SilentlyContinue).MsiInstallPath
-    if ($ExchangeServerFQDN) {
-        try {
-            if ($IgnoreSSL) {
-                SetCertificateValidationBehavior -Ignore
+    function HasWindowsVersionAmsiSupport {
+        param(
+            [Parameter(Mandatory = $true)]
+            [string]$server
+        )
+
+        $Version
+        $Version = Invoke-ScriptBlockHandler -ComputerName $server -ScriptBlock { [System.Environment]::OSVersion.Version.Major }
+        if ($Version) {
+            if ($Version -ge 10) {
+                return $true
+            } else {
+                Write-Warning "$server is not a Windows version with AMSI support."
+                return $false
             }
+        } else {
+            Write-Warning "We could not get Windows version for $server."
+            return $false
+        }
+    }
 
+    function CheckServerAMSI {
+        param(
+            [Parameter(Mandatory = $true)]
+            [string]$ExchangeServer,
+            [Parameter(Mandatory = $false)]
+            [switch]$isServer
+        )
+
+        try {
             $CookieContainer = New-Object Microsoft.PowerShell.Commands.WebRequestSession
-            $Cookie = New-Object System.Net.Cookie("X-BEResource", "a]@$($ExchangeServerFQDN):444/ecp/proxyLogon.ecp#~1941997017", "/", "$ExchangeServerFQDN")
+            $Cookie = New-Object System.Net.Cookie("X-BEResource", "a]@$($ExchangeServer):444/ecp/proxyLogon.ecp#~1941997017", "/", "$ExchangeServer")
             $CookieContainer.Cookies.Add($Cookie)
-            Invoke-WebRequest https://$ExchangeServerFQDN/ecp/x.js -Method POST -Headers @{"Host" = "$ExchangeServerFQDN" } -WebSession $CookieContainer
+            $testTime = Get-Date
+            Write-Host "Starting test at $($testTime -f "yyyy-MM-dd HH:mm:ss")"
+            if ($IgnoreSSL -and ![System.Net.ServicePointManager]::ServerCertificateValidationCallback) {
+                Enable-TrustAnyCertificateCallback
+            }
+            Invoke-WebRequest https://$ExchangeServer/ecp/x.js -Method POST -Headers @{ "Host" = "$ExchangeServer" } -WebSession $CookieContainer
         } catch [System.Net.WebException] {
-            if ($_.Exception.Message -notlike "*: (400)*") {
-                $Message = ($_.Exception.Message).ToString().Trim()
-                Write-Output $msgNewLine
-                Write-Error $Message
+            $Message = ($_.Exception.Message).ToString().Trim()
+            $currentForegroundColor = $host.ui.RawUI.ForegroundColor
+            if ($_.Exception.Message -eq "The underlying connection was closed: Could not establish trust relationship for the SSL/TLS secure channel.") {
```

Comment:
I'm not sure if this works properly for example, on systems running on a different language then en-US. Would be better to check for the status code 9 (see: https://learn.microsoft.com/en-us/dotnet/api/system.net.webexceptionstatus?view=net-7.0). 

![image](https://github.com/microsoft/CSS-Exchange/assets/40993616/44997ca8-0c58-4adb-bcfd-5cee5ddaf82e)

Reply:
This is something that comes from the original script but it looks a better solution. I will check it.

Reply:
It's the same as checking for status `9` which is `TrustFailure` (see: https://learn.microsoft.com/en-us/dotnet/api/system.net.webexceptionstatus?view=net-7.0) but it's up to you which syntax you preferer.

Reply:
yep, it is the same verification, as it is a enumeration but I think it is more clear to read but agree with you, thanks for the advise.